src/main/java/com/microsoft/azure/relay/RelayedHttpListenerResponse.java

```diff
@@ -93,9 +95,48 @@ public void close() {
 		this.closeAsync().join();
 	}
 
-	private void checkClosed() {
+	void setReadonly() {
+		this.readonly = true;
+	}
+	
+	private void checkClosedOrReadonly() {
+		String msg = null;
 		if (this.disposed) {
-			throw new RuntimeIOException("this connection has already been closed.");
+			msg = "The HTTP status code, status description, and headers cannot be modified after the output stream has already been closed.";
+		} else if (this.readonly) {
+			msg = "The HTTP status code, status description, and headers cannot be modified after writing to the output.";
+		}
+		
+		if (msg != null) {
+			throw RelayLogger.throwingException(new IllegalStateException(msg), this.context);
```

Comment:
>IllegalStateException [](start = 43, length = 21)

Just confirming, this the the correct exception type in both cases?

Reply:
How do you define "correct"? To me "IllegalStateException" makes sense for both scenarios, let me know if you feel otherwise.

---
In reply to: [277057844](https://github.com/Azure/azure-relay-java/pull/41#discussion_r277057844) [](ancestors = 277057844)

Reply:
In .NET this code throws InvalidOperationException if the operation cannot be performed at this time but the object is still viable for certain other operations. ObjectDisposedException if Close/Dispose has been called on the object.

I just didn't know if Java has any similar patterns like .NET does.

---
In reply to: [277068116](https://github.com/Azure/azure-relay-java/pull/41#discussion_r277068116) [](ancestors = 277068116,277057844)

Reply:
Using IOException for "disposed" because OutputStream throws it when attempting to write after close.

Using IllegalStateException because it's probably the closest to InvalidOperationException in C#: [stackoverflow answer](https://stackoverflow.com/questions/3165701/is-there-a-java-equivalent-of-cs-invalidoperationexception/3165733)

---
In reply to: [277070605](https://github.com/Azure/azure-relay-java/pull/41#discussion_r277070605) [](ancestors = 277070605,277068116,277057844)